man/send.man
.TH SEND %manext1% 2017-05-11 "%nmhversion%"
.
.\" %nmhwarning%
.
.SH NAME
send \- send an nmh message
.SH SYNOPSIS
.HP 5
.na
.B send
.RB [ \-help ]
.RB [ \-version ]
.RB [ \-alias
.IR aliasfile ]
.RB [ \-draft ]
.RB [ \-draftfolder
.IR +folder ]
.RB [ \-draftmessage
.IR msg ]
.RB [ \-nodraftfolder ]
.RB [ \-filter
.IR filterfile ]
.RB [ \-nofilter ]
.RB [ \-format " | " \-noformat ]
.RB [ \-forward " | " \-noforward ]
.RB [ \-mime " | " \-nomime ]
.RB [ \-msgid " | " \-nomsgid ]
.RB [ \-messageid
.IR localname " | " random ]
.RB [ \-push " | " \-nopush ]
.RB [ \-verbose " | " \-noverbose ]
.RB [ \-watch " | " \-nowatch ]
.RB [ \-mts
.IR smtp " | " sendmail/smtp " | " sendmail/pipe ]
.RB [ \-sendmail
.IR program ]
.RB [ \-server
.IR servername ]
.RB [ \-port
.IR port-name/number ]
.RB [ \-sasl ]
.RB [ \-nosasl ]
.RB [ \-saslmech
.IR mechanism ]
.RB [ \-authservice
.IR service ]
.RB [ \-snoop ]
.RB [ \-user
.IR username ]
.RB [ \-tls ]
.RB [ \-initialtls ]
.RB [ \-notls ]
.RB [ \-certverify ]
.RB [ \-nocertverify ]
.RB [ \-width
.IR columns ]
.RB [ file
\&...]
.ad
.SH DESCRIPTION
.B send
will cause each of the specified files to be delivered
to each of the destinations in the \*(lqTo:\*(rq, \*(lqcc:\*(rq,
\*(lqBcc:\*(rq, \*(lqDcc:\*(rq, and \*(lqFcc:\*(rq fields of the message.  If
.B send
is re-distributing a message, as invoked from
.BR dist ,
then the
corresponding \*(lqResent\-xxx\*(rq fields are examined instead.
.PP
By default,
.B send
uses the program
.B post
to do the actual
delivery of the messages, although this can be changed by defining the
.I postproc
profile component.  Most of the features attributed to
.B send
are actually performed by
.BR post .
.PP
Before
.B send
gives the message to
.B post
for delivery, the message is processed by
.B mhbuild
to perform any necessary MIME encoding of the outgoing message.  This
can be changed by the
.I buildmimeproc
profile component.
.B mhbuild
is invoked with the
.B \-auto
switch, so
.B mhbuild
directives are not processed by default.  See
.IR mhbuild (1)
for more information.
.PP
.B mhbuild
will scan the message draft for a header named
.IR Attach .
The draft is converted to a MIME message if one or more matches are found.
This conversion occurs before all other processing.
.IR whatnow (1)
describes the user interface for managing MIME attachments via
this mechanism.
.PP
The first part of the MIME message is the draft body if that body contains
any non-blank characters.
The body of each
.I Attach
header field is interpreted as a file name, and each file named is included as a separate
part in the MIME message.
.PP
Determination of the content MIME type inserted into the Content-Type
header for each part depends on how the
.B nmh
installation was configured.  If a program, such as
.B file
with a
.B \-\-mime
or
.B \-i
option, was found that can specify the type of a file as a MIME type
string, then that will be used.  To determine if your
.B nmh
was so configured, run
.B mhparam mimetypeproc
and see if a non-empty string is displayed.
.PP
If your
.B nmh
was not configured with a program to specify a file type as a MIME
string, then a different method is used to determine the content-type
string.  For file names with dot suffixes, the profile is scanned for a
.I mhshow-suffix-
entry for that suffix.
The content-type for the part is taken from that profile entry if a match is
found.  If a match is not found in the user profile, the mhn.defaults
profile is scanned next.
If no match is found or the file does not have a dot suffix, the content-type
is text/plain if the file contains only ASCII characters or application/octet-stream
if it contains characters outside of the ASCII range.  See
.IR mhshow (1)
for more details and example syntax.
.PP
Each attached MIME part contains a
\*(lqContent-Description\*(rq header that includes the filename, and
adds a \*(lqContent-Disposition\*(rq header.
Here is an example of MIME part headers for an attachment:
.PP
.nf
Content-Type: text/plain; name="VERSION"; charset="us-ascii"
Content-Description: VERSION
Content-Disposition: attachment; filename="VERSION"
.fi
.PP
See
.IR mhbuild (1)
for explanation of how the Content-Disposition value is selected.
.PP
If
.B \-push
is specified,
.B send
will detach itself from the user's
terminal and perform its actions in the background.  If
.BR push 'd
and the draft can't be sent, then an error message will be sent (using
the mailproc) back to the user.  If
.B \-forward
is given, then a copy
of the draft will be attached to this failure notice.  Using
.B \-push
differs from putting
.B send
in the background because the output is
trapped and analyzed by
.BR nmh .
.PP
If
.B \-verbose
is specified,
.B send
will indicate the interactions
occurring with the transport system, prior to actual delivery.
If
.B \-watch
is specified
.B send
will monitor the delivery of local
and network mail.  Hence, by specifying both switches, a large detail
of information can be gathered about each step of the message's entry
into the transport system.
.PP
The
.B \-draftfolder
.I +folder
and
.B \-draftmessage
.I msg
switches invoke
the
.B nmh
draft folder facility.  This is an advanced (and highly
useful) feature.  Consult
.IR mh-draft (5)
for more
information.
.PP
.B send
with no
.I file
argument will query whether the draft
is the intended file, whereas
.B \-draft
will suppress this question.
Once the transport system has successfully accepted custody of the
message, the file will be renamed with a site-dependent prefix
(usually a comma), which allows
it to be retrieved until the next draft message is sent.  If there are
errors in the formatting of the message,
.B send
will abort with a
(hopefully) helpful error message.
.PP
If a \*(lqBcc:\*(rq field is encountered, its addresses will be used for
delivery, and the \*(lqBcc:\*(rq field will be removed from the message
sent to sighted recipients.  The blind recipients will receive an entirely
new message with a minimal set of headers. The body of this new message
will contain a copy of the message sent to the sighted recipients, either
marked up with the indicator text "Blind-Carbon-Copy" or encapsulated
as a MIME digest.
.PP
If a \*(lqDcc:\*(rq field is encountered and the
.B sendmail/pipe
mail transport method is not in use, its addresses will be used for
delivery, and the \*(lqDcc:\*(rq field will be removed from the message.  The
blind recipients will receive exactly the same message as the sighted
recipients.  *WARNING* Recipients listed in the \*(lqDcc:\*(rq field receive no
explicit indication that they have received a \*(lqblind copy\*(rq.
This can cause blind recipients to
inadvertently reply to all of the sighted recipients of the
original message, revealing that they received a blind copy.
On the other hand, since a normal reply to a message sent
via a \*(lqBcc:\*(rq field
will generate a reply only to the sender of the original message,
it takes extra effort in most mailers to reply to the included
message, and so would usually only be done deliberately, rather
than by accident.
.PP
If the
.B sendmail/pipe
mail transport method is used, then messages containing
a \*(lqDcc:\*(rq field are rejected.

.PP
If
.B \-filter
.I filterfile
is specified, then this copy is filtered
(re-formatted) by
.B mhl
prior to being sent to the blind recipients.
Alternately, if you specify the
.B -mime
switch, then
.B send
will
use the MIME rules for encapsulation.
.PP
Prior to sending the message, the \*(lqDate:\ now\*(rq field will be appended to the headers in the message.
If
.B \-msgid
is specified, then a \*(lqMessage\-ID:\*(rq field will also
be added to the message.
.PP
The
.B \-messageid
switch selects the style used for the part appearing after the @
in \*(lqMessage\-ID:\*(rq, \*(lqResent\-Message\-ID:\*(rq, and
\*(lqContent\-ID:\*(rq header fields.  The two acceptable options are
.B localname
(which is the default),
and
.BR random .
With
.BR localname ,
the local hostname is used.  With
.BR random ,
a random sequence of characters is used instead.  Note that the
.B \-msgid
switch must be enabled for this switch to have any effect.
.PP
If
.B send
is re-distributing a message (when invoked by
.BR dist ),
then \*(lqResent\-\*(rq will be prepended to each of these
fields: \*(lqFrom:\*(rq, \*(lqDate:\*(rq, and \*(lqMessage\-ID:\*(rq.
.PP
A \*(lqFrom:\*(rq field is required for all outgoing messages.  Multiple
addresses are permitted in the \*(lqFrom:\*(rq field, but a \*(lqSender:\*(rq
field is required in this case.  Otherwise a \*(lqSender:\*(rq field
is optional.
.PP
If a message with multiple \*(lqFrom:\*(rq
addresses does
.I not
include a \*(lqSender:\*(rq field but does include an \*(lqEnvelope\-From:\*(rq
field, the \*(lqEnvelope\-From:\*(rq field will be used to construct
a \*(lqSender:\*(rq field.
.PP
When using SMTP for mail submission, the envelope-from used for the SMTP
transaction is derived from the \*(lqEnvelope\-From:\*(rq field.
If no \*(lqEnvelope\-From:\*(rq field is present, the \*(lqSender:\*(rq
field is used.  If neither the \*(lqEnvelope\-From:\*(rq nor the
\*(lqSender:\*(rq field is present, the \*(lqFrom:\*(rq field is used.
When \*(lqEnvelope\-From:\*(rq appears in a message
it will be removed from the final outgoing message.
.PP
By using the
.B \-format
switch, each of the entries in the \*(lqTo:\*(rq
and \*(lqcc:\*(rq fields will be replaced with \*(lqstandard\*(rq
format entries.  This standard format is designed to be usable by all
of the message handlers on the various systems around the Internet.
If
.B \-noformat
is given, then headers are output exactly as they appear
in the message draft.
.PP
If an \*(lqFcc:\ folder\*(rq is encountered, the message will be copied
to the specified folder for the sender in the format in which it will
appear to any non\-Bcc receivers of the message.  That is, it will have
the appended fields and field reformatting.  The \*(lqFcc:\*(rq fields
will be removed from all outgoing copies of the message.
.PP
Beware that if an \*(lqFcc:\*(rq with one or more folders is present
but none of the folders exist, and the default
.I fileproc
and
.I postproc
are in use, then
.B refile
will prompt the user to create the folder(s) if
.B \-push
is not specified.  If all responses are negative, or creation of each folder
fails, or
.B \-push
is specified, the message will not be copied to any folder and will be
removed by
.BR post .
With the default
.B refile
switches, the message draft will be renamed according to the specification
of its
.B \-nolink
switch.
.PP
By using the
.B \-width
.I columns
switch, the user can direct
.B send
as to how long it should make header lines containing addresses.
.PP
The mail transport system default is provided in
.I %nmhetcdir%/mts.conf
but can be overridden here with the
.B \-mts
switch.
.PP
If nmh is using
.B sendmail/pipe
or
.B sendmail/smtp
as its mail transport system, the
.B \-sendmail
switch can be used to override the default
.B sendmail
program.
.PP
If nmh is using the SMTP MTA, the
.B \-server
and the
.B \-port
switches can be used to override the default mail server (defined by the
.I %nmhetcdir%/mts.conf
.I servers
entry).  The
.B \-snoop
switch can be used to view the SMTP transaction.  (Beware that the
SMTP transaction may contain authentication information either in
plaintext or easily decoded base64.)  If
.B \-sasl \-saslmech xoauth2
is used, the HTTP transaction is also shown.
.PP
If
.B nmh
has been compiled with SASL support, the
.B \-sasl
and
.B \-nosasl
switches will enable and disable
the use of SASL authentication with the SMTP MTA.  Depending on the
SASL mechanism used, this may require an additional password prompt from the
user (but the
.I netrc
file can be used to store this password, as described in
.IR mh-profile (5).
The
.B \-saslmech
switch can be used to select a particular SASL mechanism,
and the
.B \-user
switch can be used to select a authorization userid to provide to SASL
other than the default.  The credentials profile entry in
.IR mh\-profile (5)
describes the ways to supply a username and
password.
.PP
If SASL authentication is successful,
nmh
will attempt to negotiate a security layer for session encryption.
Encrypted data is labelled with `(encrypted)' and `(decrypted)' when
viewing the SMTP transaction with the
.B \-snoop
switch; see
.IR post (8)'s
description of
.B \-snoop
for its other features.
.PP
If
.B nmh
has been compiled with OAuth support, the
.B \-sasl
and
.B \-saslmech xoauth2
switches will enable OAuth authentication.  The
.B \-user
switch must be used, and the
.I username
must be an email address the user has for the service, which must
be specified with the
.B \-authservice
.I service
switch.  Before using OAuth authentication, the user must authorize nmh by
running
.B mhlogin
and grant authorization to that account.  See
.IR mhlogin (1)
for more details.
.PP
If
.B nmh
has been compiled with TLS support, the
.B \-tls
and
.B \-initialtls
switches will require the negotiation of TLS when
connecting to the SMTP MTA.  The
.B \-tls
switch will negotiate TLS as part of the normal SMTP protocol
using the STARTTLS command.  The
.B \-initialtls
will negotiate TLS immediately after the connection has
taken place, before any SMTP commands are sent or received.  Encrypted data
is labelled with `(tls-encrypted)' and
`(tls-decrypted)' when viewing the SMTP transaction with the
.B \-snoop
switch; see
.IR post (8)'s
description of
.B \-snoop
for its other features.
The
.B \-notls
switch will disable all attempts to negotiate TLS.
.PP
If port 465 is specified and none of the TLS switches were enabled,
.B \-initialtls
will be implied if TLS support was compiled in.  Though port 465 for
SMTPS (SMTP over SSL) was deregistered by IANA in 1998, it is still
used for that service.
.PP
When using TLS the default is to verify the remote certificate and SubjectName
against the local trusted certificate store.  This can be controlled by
the
.B \-certverify
and
.B \-nocertverify
switches.  See your OpenSSL documentation for more information on certificate
verification.
.PP
The files specified by the profile entry \*(lqAliasfile:\*(rq and any
additional alias files given by the
.B \-alias
.I aliasfile
switch will be
read (more than one file, each preceded by
.BR \-alias ,
can be named).
See
.IR mh\-alias (5)
for more information.
.SS Selection based on sender address:  sendfrom
One or more
.I sendfrom
profile components can be used to select a mail server address, mail server
port, or any other switch that can be supplied to
.BR post .
It works by first looking at the sender address and domain name in the
message draft, as described below.
It then looks for a corresponding profile entry, which contains the
.B post
switches.
To enable, add profile entries of the form:
.PP
.RS 5
.RI sendfrom- "address/domain name" : " post switches"
.RE
.PP
The email address is extracted from the Envelope-From:  header, if not blank,
the Sender: header, or the From: header line in the message draft.
Multiple profile entries, with different email addresses or domain names, are
supported.
This allows different switches to
.BR post ,
such as -user, to be associated with different email addresses.
If a domain name is used, it matches all users in that domain.
.PP
Here is an example profile entry using OAuth for an account hosted by gmail:
.PP
.nf
.RS 5
sendfrom-gmail_address@example.com: -sasl -saslmech xoauth2
.RS 5
-authservice gmail -tls -server smtp.gmail.com
-user gmail_login@example.com
.RE
.RE
.fi
.PP
(Indentation indicates a continued line, as supported in MH profiles.)
The username need not be the same as the sender address, which was extracted
from the appropriate header line as noted above.
.PP
Here are example profile entries that use an nmh credentials file:
.PP
.nf
.RS 5
credentials: file:nmhcreds
sendfrom-sendgrid_address@example.com: -sasl -tls
.RS 5
-server smtp.sendgrid.net
.RE
sendfrom-outbound.att.net: -sasl -initialtls
.RS 5
-server outbound.att.net -port 465
.RE
sendfrom-fastmail.com: -initialtls -sasl -saslmech LOGIN
.RS 5
-server smtps-proxy.messagingengine.com -port 80
.RE
.RE
.fi
.PP
where nmhcreds is in the user's nmh directory (from the Path profile component)
and contains:
.PP
.nf
.RS 5
machine smtp.sendgrid.net
.RS 5
login sendgrid_login@example.com
password ********
.RE
machine outbound.att.net
.RS 5
login att_login@example.com
password ********
.RE
machine smtps-proxy.messagingengine.com
.RS 5
login fastmail_login@example.com
password ********
.RE
.RE
.fi
.PP
For more information on authentication to mail servers, see
.IR mhlogin (1)
for OAuth services, and
.IR mh-profile (5)
for login credentials.
.PP
.SH FILES
.fc ^ ~
.nf
.ta \w'%nmhetcdir%/ExtraBigFileName  'u
^$HOME/.mh_profile~^The user profile
.fi
.SH "PROFILE COMPONENTS"
.fc ^ ~
.nf
.ta 2.4i
.ta \w'ExtraBigProfileName  'u
^Path:~^To determine the user's nmh directory
^Draft\-Folder:~^To find the default draft-folder
^Aliasfile:~^For a default alias file
^Signature:~^To determine the user's mail signature
^mailproc:~^Program to post failure notices
^postproc:~^Program to post the message
^sendfrom-address:~^Switches to post for sender address
^sendfrom-domain:~^Switches to post for sender domain name
.fi
.SH "SEE ALSO"
.IR comp (1),
.IR dist (1),
.IR file (1),
.IR forw (1),
.IR mhbuild (1),
.IR mhparam (1),
.IR mhlogin (1),
.IR refile (1),
.IR repl (1),
.IR whatnow (1),
.IR mh\-alias (5),
.IR mh\-profile (5),
.IR mh\-tailor (5),
.IR post (8)
.SH DEFAULTS
.nf
.RB ` file "' defaults to <mh-dir>/draft"
.RB ` \-alias "' defaults to %nmhetcdir%/MailAliases"
.RB ` \-nodraftfolder '
.RB ` \-nofilter '
.RB ` \-format '
.RB ` \-forward '
.RB ` \-nomime '
.RB ` \-nomsgid '
.RB ` "\-messageid\ localname" '
.RB ` \-nopush '
.RB ` \-noverbose '
.RB ` \-nowatch '
.RB ` "\-width\ 72" '
.RB ` \-certverify '
.fi
.SH CONTEXT
None
.SH BUGS
Under some configurations, it is not possible to monitor the mail delivery
transaction;
.B \-watch
is a no-op on those systems.


NEWS
# NEWS - Release notes for nmh 1.7
#

Welcome to nmh, the new version of the classic MH mail handling system.
It's been over three years since the last release of nmh, and there have
been a number of significant changes since the last release.  Long-time
MH and nmh uses should read careful the NOTEABLE CHANGES section, as there
are some significant changes to nmh behavior.  Otherwise, please see the
README and INSTALL files for help on getting started with nmh.

For news of future releases, subscribe to the low-volume
    https://lists.nongnu.org/mailman/listinfo/nmh-announce

---------------
NOTABLE CHANGES
---------------

The largest notable changes in the 1.7 release are:

- Complete unification of network security support.  All network protocols
  (currently, POP and SMTP) have been refactored to use a common set of
  security routines.  This means all protocols support all SASL mechanisms
  (via the Cyrus-SASL library) and TLS.  TLS support has been strengthened
  to perform certificate name validation and to require TLS 1.1 as a
  minimum protocol.  Also, all protocols can make use of the OAuth2/XOAUTH
  SASL mechanism, which is supported by Gmail.
- send(1) now supports adding switches to post(8) based on the address or
  domain of the email address in the From: header; this more easily allows
  users to support multiple identities.
- A generic facility for passing arguments to filter programs in repl(1)
  by use of the -convertargs switch.
- Native support for the manipulation of iCalendar requests; see mhical(1)
  for more details.

------------
NEW FEATURES
------------

The following are new features for the 1.7 release of nmh:

- When building from source, configure will derive ${prefix} from an existing
  nmh installation if it finds one in your $PATH.
- Added welcome message when nmh detects that its version changed.
- The default locations for configuration files and support binaries
  have been changed.  Configuration files now install into ${sysconfdir}/nmh,
  and support binaries are placed in ${libexecdir}/nmh.  If you are upgrading
  an existing installation you should look for old configuration files in
  ${sysconfdir} and merge any local customizations into the new files in
  ${sysconfdir}/nmh, then remove the old files.  ${libdir} will also contain
  obsolete support programs that should be removed.
- All TLS connections now perform certificate validation (including hostname
  matching) by default; can be disabled on a per-application basis.
- post now defaults to port 587 on 'smtp' message submission.
- A value of 0 for the width switch of scan(1), inc(1), ap(1), dp(1),
  fmttest(1), and mhl(1) now means as many characters as the format
  engine can produce [Bug #15274].  That amount is limited by internal
  buffers.
- If a component has trailing whitespace, e.g., body:component="> ",
  mhl now trims that whitespace off when filtering blank text lines.
- An "rtrim" flag has been added to mhl to remove any trailing
  whitespace from filtered text lines.  A corresponding "nortrim" flag
  has also been added.
- Added getmymbox and getmyaddr mh-format(5) function escapes.
- New -[no]changecur, -fixtype, -decodetypes, and -[no]crlflinebreaks switches
  have been added to mhfixmsg(1).
- mhfixmsg now removes an extraneous trailing semicolon from header
  parameter lists.
- Added -convertargs switch to repl(1), to pass arguments to programs
  specified in the user's profile or mhn.defaults to convert message
  content.
- Added mhical(1), to display, reply to, and cancel iCalendar (RFC 5545)
  event requests.
- Added multiply mh-format(5) function.
- "mhparam bindir" prints the path to the directory containing the public
  executables (${bindir}).
- New "-prefer" switch for mhshow (and mhlist and mhshow), to allow specifying
  the preferred content types to show, if present in a multipart alternative.
- mh-format now has %(kilo) and %(kibi) functions, to allow printing
  numbers with SI or IEC quantities, e.g. "10K", "2.3Mi".
- Support for the -sendmail flag to send/post to change the sendmail
  binary when using the sendmail/pipe MTS.
- Added support to send(1) to specify switches to post(1) based on address or
  domain name in From: header line in message draft.
- post(8) -snoop now attempts to decode base64-encoded SMTP traffic.
- folder(1) -nocreate now prints a warning message for a non-existent folder.
- mhfixmsg(1) now allows -decodetext binary, though 8bit is still the default.
- inc(1) and msgchk(1) now support TLS encryption natively.
- All network protocols support the XOAUTH authentication mechanism.
- Support for SMTPUTF8 (RFC 6531) has been added.  mhshow(1) already supported
  RFC 6532, assuming all 8-bit message header field bodies are UTF-8 and use
  of a UTF-8 locale.
- mhfixmsg now replaces RFC 2047 encoding with RFC 2231 encoding of name and
  filename parameters in Content-Type and Content-Disposition headers,
  respectively.
- If a message body contains 8-bit bytes, post(8) uses SMTP 8BITMIME if the
  server supports it.  If not, post fails with a message to the user to
  encode the message for 7-bit transport.
- Fewer lseek(2)s will be used when reading headers in the common case.
- ./configure's --enable-debug has been removed;  it did nothing.
- configure now defaults to enabling each of TLS and Cyrus SASL if the
  necessary headers and libraries are found.
- Moved build_nmh to top-level directory.
- Better error reporting for connections to network services.

---------
BUG FIXES
---------

- The format scanner no longer subtracts 1 from the width.  This has the
  effect of no longer counting the trailing newline in the output of
  scan(1), inc(1), and the other programs that rely on the format scanner.
- The first character of some very short (less than 4 characters) message
  bodies is no longer dropped.
- Single-character headers can be reliably formatted, etc., instead of
  apparently being missing.

docs/historical/2.9BSD/cmds/send.c
#ifndef lint
static char sccsid[] = "@(#)send.c	4.2 2/23/83";
#endif

#include "mh.h"
#include <stdio.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <strings.h>
#include <signal.h>

char   *anoyes[];       /* Std no/yes gans array        */

char   *vec[20];
int     vecp;

struct swit switches[] = {
	"debug",         -1,      /* 0 */
	"draft",          0,      /* 1 */
	"format",         0,      /* 2 */
	"noformat",       0,      /* 3 */
	"msgid",          0,      /* 4 */
	"nomsgid",        0,      /* 5 */
	"verbose",        0,      /* 6 */
	"noverbose",      0,      /* 7 */
	"help",           4,      /* 8 */
	0,                0
};

int     debug;

main(argc, argv)
char *argv[];
{
	register char *drft, *cp;
	register int i;
	int status, pid;
	struct stat stbuf;
	char **ap;
	char *arguments[50], **argp;

#ifdef NEWS
	m_news();
#endif
	drft = 0;
	vec[vecp++] = "mh_deliver";
	cp = r1bindex(argv[0], '/');
	if((cp = m_find(cp)) != NULL) {
		ap = brkstring(cp = getcpy(cp), " ", "\n");
		ap = copyip(ap, arguments);
	} else
		ap = arguments;
	copyip(argv+1, ap);
	argp = arguments;
	while(cp = *argp++) {
		if(*cp == '-')
			switch(smatch(++cp, switches)) {
			case -2:ambigsw(cp, switches);       /* ambiguous */
				goto leave;
							     /* unknown */
			case -1:fprintf(stderr, "send: -%s unknown\n", cp);
				goto leave;
							     /* -draft */
			case 1: vec[vecp++] = drft = m_maildir(draft);
				continue;
			case 0: debug++;
			case 2: case 3: case 4:
			case 5: case 6: case 7:
				vec[vecp++] = --cp;
				continue;
			case 8: help("send [file]   [switches]",
				     switches);
				goto leave;
			}
		if(drft) {
			fprintf(stderr, "Send: Only one message at a time.\n");
			done(1);
		}
		vec[vecp++] = drft = cp;
	}
	if(!drft) {
		drft = m_maildir(draft);
		if(stat(drft, &stbuf) == -1) {
			fprintf(stderr, "Draft file: %s doesn't exist.\n", drft);
			done(1);
		}
		cp = concat("Use \"", drft, "\"? ", 0);
		if(!gans(cp, anoyes))
			done(0);
		vec[vecp++] = drft;
	} else {
		if(stat(drft, &stbuf) == -1) {
			fprintf(stderr, "Draft file: %s doesn't exist.\n", drft);
			done(1);
		}
	}
	m_update();
	vec[vecp] = 0;

	while((pid = fork()) == -1) {
		fprintf("Waiting for a fork\n");
		sleep(2);
	}
	if(pid == 0) {
		execv(mh_deliver, vec);
		perror(mh_deliver);
		done(1);
	}
	signal(SIGINT, SIG_IGN);
	signal(SIGQUIT, SIG_IGN);
	while((i = wait(&status)) != -1 && i != pid) ;
	if(status == 0 && !debug)
		backup(drft);

leave:  ;
/***    m_update();     ***/
}


backup(file)
char *file;
{
	char buf[128];
	register char *cp;
	extern char *rindex();

	buf[0] = 0;
	if(cp = rindex(file, '/'))
		sprintf(buf, "%.*s", (++cp)-file, file);
	else
		cp = file;
#ifdef UCB
	strcat(buf, "#");
#else
	strcat(buf, ",");
#endif
	strcat(buf, cp);
	unlink(buf);
	if(link(file, buf) < 0 || unlink(file) < 0) {
		fprintf(stderr, "Send: Backup rename failure ");
		perror(buf);
		done(1);
	}
}
